**Discretización**

**Edad**

In [ ]:
x     y
25    1 (reclaman)
35    0
41    0
28    1 (reclaman)
29    1 (reclaman)
.
.
.
10 k casos

**Naive Bayes**: basado en la **probabilidad condicional**

In [ ]:
Pr(x=¿?|y=1)

In [ ]:
Pr(x=20|y=1)=0.1
Pr(x=21|y=1)=0.01
Pr(x=22|y=1)=0.1
.
.
.
Pr(x=58|y=1)=0.005

**¿Cómo discretizo una variable?**

In [ ]:
1 = 20-30
2 = 31-45
3 = +45

In [ ]:
Pr(x=1|y=1)=0.25
Pr(x=2|y=1)=0.50
Pr(x=3|y=1)=0.25

**Árboles de decisión**

**impureza**: error del árbol (entropia o gini)

In [ ]:
a=6
d=4

impureza inicial

**Nota**: las variables que se agregan al modelo de árbol son discretas, si no lo son el modelo de árbol las discretiza utilizando entropia.

**Discretización por intervalos de igual amplitud**

In [29]:
from sklearn.preprocessing import KBinsDiscretizer
import math as m
import pandas as pd
import numpy as np

In [30]:
data = pd.read_csv("D:/Python/1. Nivel I/7/datos/data.csv")

In [31]:
data.head()

Unnamed: 0      rsrp  redireccion       cqi  incidentes
0           0 -116.5000  90890.42778  8.138370           0
1           1 -115.1875  91193.46144  9.140266           0
2           2 -119.1250  91877.50918  7.808917           0
3           3 -111.8125  92049.63360  8.878271           0
4           4 -115.1875  92275.58028  9.051622           0

In [32]:
data.drop(columns="Unnamed: 0",axis=1, inplace=True)

In [33]:
data.head()

rsrp  redireccion       cqi  incidentes
0 -116.5000  90890.42778  8.138370           0
1 -115.1875  91193.46144  9.140266           0
2 -119.1250  91877.50918  7.808917           0
3 -111.8125  92049.63360  8.878271           0
4 -115.1875  92275.58028  9.051622           0

In [9]:
data["cqi"].min()

4.0703296661377

In [10]:
data["cqi"].max()

11.62320369482

In [11]:
rango = data["cqi"].max() - data["cqi"].min(); print(rango)

7.552874028682301


In [16]:
(4*(rango/4)) + 4.0703296661377

11.62320369482

In [17]:
rango/4

1.8882185071705753

In [ ]:
4.0703296661377 - 5.958548173308275  ---> categoría 1
5.958548173308275 - 7.846766680478851 --> categoría 2
7.846766680478851 - 9.734985187649425 --> categoría 3
9.734985187649425 - 11.62320369482 -->    categoría 4

**¿Cómo determino el número de categorías?**

In [34]:
sturges = round(1+ m.log2(len(data)),0); print(sturges)

12.0


In [19]:
rango/sturges

0.6294061690568584

**función KBinsDiscretizer()**

**Nota**: este proceso de discretización está afectado por la presencia de valores atípicos. No es recomendable aplicarlo cuando ustedes noten la presencia de valores atípicos

In [35]:
amplitud =  KBinsDiscretizer(n_bins=12, #número de categorías de la variable
                            encode="ordinal", #son de naturaleza ordinal
                            strategy="uniform") #uniform= discretización por intervalos de igual amplitud

In [36]:
data["cqi_inter_amplitud"]= amplitud.fit_transform(data[["cqi"]])

In [37]:
data.head()

rsrp  redireccion       cqi  incidentes  cqi_inter_amplitud
0 -116.5000  90890.42778  8.138370           0                 6.0
1 -115.1875  91193.46144  9.140266           0                 8.0
2 -119.1250  91877.50918  7.808917           0                 5.0
3 -111.8125  92049.63360  8.878271           0                 7.0
4 -115.1875  92275.58028  9.051622           0                 7.0

**Discretización por intervalos de igual frecuencia o discretización cuantil**

In [23]:
datan = [4,3,2,6,8,2,5,6]

**Paso 1**: de forma ascendente los valores de la variable

In [24]:
datan = [2,2,3,4,5,6,6,8]

**Paso 2: cuartiles [0.25,0.50,0.75,1]**

In [ ]:
categoría 1 = [2,2]
categoría 2 = [3,4]
categoría 3 = [5,6,6]
categoría 4 = [8]

In [38]:
cuartil =  KBinsDiscretizer(n_bins=4, #número de categorías de la variable
                            encode="ordinal", #son de naturaleza ordinal
                            strategy="quantile") #quantile= discretización por intervalos de igual frecuencia

In [39]:
data["cqi_cuartil"]= cuartil.fit_transform(data[["cqi"]])

In [40]:
data.head()

rsrp  redireccion       cqi  incidentes  cqi_inter_amplitud  \
0 -116.5000  90890.42778  8.138370           0                 6.0   
1 -115.1875  91193.46144  9.140266           0                 8.0   
2 -119.1250  91877.50918  7.808917           0                 5.0   
3 -111.8125  92049.63360  8.878271           0                 7.0   
4 -115.1875  92275.58028  9.051622           0                 7.0   

   cqi_cuartil  
0          1.0  
1          2.0  
2          0.0  
3          2.0  
4          2.0

**Discretización por KMeans**

<center><img src="https://miro.medium.com/max/963/1*Qu9vtRFFrQ1_4jj4Nu__BQ.jpeg" width="700"></center>

In [41]:
datos = [23,25,22,29,36,41,38,18]

In [42]:
c1= 26
c2= 32

In [43]:
data_cl = [1,1,1,2,2,2,2,1]

In [ ]:
c1= 22
c2= 36

In [45]:
data_cl = [1,1,1,1,2,2,2,1]

In [ ]:
c1= 23.4
c2= 38.33

In [48]:
data_cl = [1,1,1,1,2,2,2,1]

**El algoritmo se detiene porque las observaciones dejan de reasignarse a otro cluster**

In [52]:
df = pd.DataFrame(datos,columns=["edad"])

In [53]:
df

edad
0    23
1    25
2    22
3    29
4    36
5    41
6    38
7    18

In [54]:
kmeans =  KBinsDiscretizer(n_bins=2, #número de categorías de la variable
                            encode="ordinal", #son de naturaleza ordinal
                            strategy="kmeans") #kmeans= discretización por k-medias

In [57]:
df["edad_kmeans"]=kmeans.fit_transform(df[["edad"]])+1

In [58]:
df

edad  edad_kmeans
0    23          1.0
1    25          1.0
2    22          1.0
3    29          1.0
4    36          2.0
5    41          2.0
6    38          2.0
7    18          1.0

In [ ]:
data_cl = [1,1,1,1,2,2,2,1]

**Nota**: K-means está afecta a la presencia de valores atípicos

In [59]:
kmeans =  KBinsDiscretizer(n_bins=2, #número de categorías de la variable
                            encode="ordinal", #son de naturaleza ordinal
                            strategy="kmeans") #kmeans= discretización por k-medias

In [60]:
data["cqi_kmeans"]= kmeans.fit_transform(data[["cqi"]])

In [61]:
data.head()

rsrp  redireccion       cqi  incidentes  cqi_inter_amplitud  \
0 -116.5000  90890.42778  8.138370           0                 6.0   
1 -115.1875  91193.46144  9.140266           0                 8.0   
2 -119.1250  91877.50918  7.808917           0                 5.0   
3 -111.8125  92049.63360  8.878271           0                 7.0   
4 -115.1875  92275.58028  9.051622           0                 7.0   

   cqi_cuartil  cqi_kmeans  
0          1.0         0.0  
1          2.0         1.0  
2          0.0         0.0  
3          2.0         1.0  
4          2.0         1.0

**Discretización por entropia o discretización basada en árboles**

- basada en un modelo de aprendizaje : árboles de decisión
- la discretización de la variable x se realizará en función a la variable y  (target)

In [ ]:
x     y
25    1 (reclaman)
35    0
41    0
28    1 (reclaman)
29    1 (reclaman)
.
.
.
10 k casos


<center><img src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTOGIWR9duZ9cM0Rv8mi6WaN0bEpi11oI81Gg&usqp=CAU" width="400"></center>

**Paso 1**: calcular la impureza inicial del árbol por entropia

In [62]:
r=4000
nr=6000
t=10000

In [67]:
impureza=abs(((r/t)*m.log(r/t))+((nr/t)*m.log(nr/t)));print(impureza)

0.6730116670092565


**Se va discretizar la variable x utilizando diferentes puntos de corte (puntos de discretización) y ganará el que reduzca la impureza del árbol**

La impureza inicial del árbol es 0.673 y la impureza final usando el punto de corte "30 años" es 0.672

La impureza inicial del árbol es 0.673 y la impureza final usando el punto de corte "38 años" es 0.35

La impureza inicial del árbol es 0.673 y la impureza final usando el punto de corte "40 años" es 0.27

**Paso 3** seleccionamos el punto de corte de discretización

**Nota** se selecciona el punto de corte que reduzca la impureza inicial del árbol de forma más significativa

**Nodo 1** de 18 a 30 años

In [77]:
t=3500
r=1500
nr=2000

In [78]:
impureza1=abs(((r/t)*m.log(r/t))+((nr/t)*m.log(nr/t)));print(impureza1)

0.6829081047004717


In [79]:
impureza1*0.35

0.23901783664516507

**Nodo 2** de 31 a más años

In [73]:
t=6500
r=2500
nr=4000

In [75]:
impureza2=abs(((r/t)*m.log(r/t))+((nr/t)*m.log(nr/t)));print(impureza2)

0.666278442414676


In [76]:
impureza2*0.65

0.43308098756953944

**calculo de la impureza final de árbol**

In [80]:
suma_i=(impureza1*0.35)+(impureza2*0.65);print(suma_i)

0.6720988242147046


In [81]:
import os
os.getcwd()

'C:\\Users\\SOCIAL DATA'

**Nota**: descargar **MDLP.py** & **Entropy.py** del Classrom y ubicarlos en el directorio de trabajo actual

In [82]:
from MDLP import MDLP_Discretizer

In [84]:
data.head()

rsrp  redireccion       cqi  incidentes  cqi_inter_amplitud  \
0 -116.5000  90890.42778  8.138370           0                 6.0   
1 -115.1875  91193.46144  9.140266           0                 8.0   
2 -119.1250  91877.50918  7.808917           0                 5.0   
3 -111.8125  92049.63360  8.878271           0                 7.0   
4 -115.1875  92275.58028  9.051622           0                 7.0   

   cqi_cuartil  cqi_kmeans  
0          1.0         0.0  
1          2.0         1.0  
2          0.0         0.0  
3          2.0         1.0  
4          2.0         1.0

In [85]:
x=data.iloc[:,0:3].values
y=data.iloc[:,3].values

In [86]:
ba=[2]

In [87]:
n_predictores = np.array(ba)

In [88]:
n_predictores

array([2])

In [89]:
discretizer = MDLP_Discretizer(features=n_predictores)

In [90]:
discretizer.fit(x,y)

In [91]:
data["cqi_entropia"]=discretizer.transform(x)[:,2]

In [92]:
data.head()

rsrp  redireccion       cqi  incidentes  cqi_inter_amplitud  \
0 -116.5000  90890.42778  8.138370           0                 6.0   
1 -115.1875  91193.46144  9.140266           0                 8.0   
2 -119.1250  91877.50918  7.808917           0                 5.0   
3 -111.8125  92049.63360  8.878271           0                 7.0   
4 -115.1875  92275.58028  9.051622           0                 7.0   

   cqi_cuartil  cqi_kmeans  cqi_entropia  
0          1.0         0.0           0.0  
1          2.0         1.0           1.0  
2          0.0         0.0           0.0  
3          2.0         1.0           1.0  
4          2.0         1.0           1.0

In [93]:
#############################################

**Escalamiento de variables**

**Ciertas técnicas (modelos) requieren que los datos estén en una misma escala de medida**

- Algoritmos de agrupamiento de observaciones k-means & AGNES (clúster jerárquico aglomerativo)
- técnicas de reducción de dimensionalidad: PCA & FA

In [94]:
import os
import pandas as pd
from sklearn.model_selection import train_test_split

In [95]:
os.chdir("D:/Python/1. Nivel I/9/datos")

In [96]:
df = pd.read_csv("data_credit.csv",
                sep=",",
                encoding="ISO-8859-1")

In [97]:
df.head()

duration  chist  reason    camt  telephne  lenemp  instrate  perstat  \
0       6.0    5.0     4.0  1169.0       0.0     4.0       4.0      3.0   
1      48.0    3.0     4.0  5951.0       1.0     2.0       2.0      2.0   
2      12.0    5.0     7.0  2096.0       1.0     3.0       2.0      3.0   
3      42.0    3.0     3.0  7882.0       1.0     3.0       2.0      3.0   
4      24.0    4.0     1.0  4870.0       1.0     2.0       3.0      3.0   

   residlen  prpownr   age  housng  numcred  emptype  numliab  customer  
0       4.0      1.0  67.0     2.0      2.0      3.0      1.0       1.0  
1       2.0      1.0  22.0     2.0      1.0      3.0      1.0       2.0  
2       3.0      1.0  49.0     2.0      1.0      2.0      2.0       1.0  
3       4.0      2.0  45.0     3.0      1.0      3.0      2.0       1.0  
4       4.0      4.0  53.0     3.0      2.0      3.0      2.0       2.0

**Particionamiento**

In [98]:
x=df.iloc[:,0:15].values
y=df.iloc[:,15].values

In [99]:
x_train, x_test, y_train, y_test = train_test_split(x,y,
                                                   test_size=0.3,
                                                   random_state=2023)

**Normalización**

transformación de los datos es la técnica que busca que los datos tengan un comportamiento normal

In [100]:
datos = [23,21,18,32,30,26]

In [101]:
min_data=18
max_data=32

In [105]:
len(datos)

6

In [110]:
datos_norm=[]

In [111]:
for i in range (0,len(datos)):
    d=(datos[i]-min_data)/(max_data-min_data)
    datos_norm.append(d)

In [121]:
df_caso=pd.DataFrame({"datos":datos,"datos_norm":datos_norm})

In [122]:
df_caso

datos  datos_norm
0     23    0.357143
1     21    0.214286
2     18    0.000000
3     32    1.000000
4     30    0.857143
5     26    0.571429

In [114]:
from sklearn.preprocessing import MinMaxScaler

In [115]:
minmax=MinMaxScaler()

In [116]:
x_train_norm= minmax.fit_transform(x_train)

**Estandarización**

Basada en la media y desviación estándar, mantiene el comportamiento original de los datos.

In [125]:
datos = [23,21,18,32,30,26]

In [123]:
df_caso["datos"].mean()

25.0

In [124]:
np.std(df_caso["datos"])

4.898979485566356

In [127]:
datos_std=[]

In [128]:
for i in range (0,len(datos)):
    d=(datos[i]-df_caso["datos"].mean())/(np.std(df_caso["datos"]))
    datos_std.append(d)

In [130]:
df_caso["datos_std"]=datos_std

In [131]:
df_caso

datos  datos_norm  datos_std
0     23    0.357143  -0.408248
1     21    0.214286  -0.816497
2     18    0.000000  -1.428869
3     32    1.000000   1.428869
4     30    0.857143   1.020621
5     26    0.571429   0.204124

In [132]:
from sklearn.preprocessing import StandardScaler

In [133]:
std=StandardScaler()

In [134]:
x_train_std=std.fit_transform(x_train)

**Reducción de dimensionalidad**

1. problema multicolinealidad (predictores están correlacionadas)

2. solución: aplicar técnicas de reducción de dimensionalidad (PCA o FA). Estas técnicas decorrelacionan las variables creando componentes o factores

**Paso 1**: demostrar que existe multicolinealidad

In [135]:
df_train = pd.DataFrame(x_train,
                       columns=df.columns.tolist()[0:15])

In [136]:
df_train.corr(method="pearson")

duration     chist    reason      camt  telephne    lenemp  \
duration  1.000000 -0.090075  0.173382  0.639104 -0.187380  0.050090   
chist    -0.090075  1.000000 -0.071030 -0.060820 -0.077742  0.173516   
reason    0.173382 -0.071030  1.000000  0.219877 -0.134214  0.066568   
camt      0.639104 -0.060820  0.219877  1.000000 -0.274351  0.060297   
telephne -0.187380 -0.077742 -0.134214 -0.274351  1.000000 -0.154230   
lenemp    0.050090  0.173516  0.066568  0.060297 -0.154230  1.000000   
instrate  0.073543  0.054341 -0.000763 -0.264340 -0.022830  0.054605   
perstat   0.027830  0.028894 -0.015217 -0.029019 -0.022277  0.080002   
residlen  0.020677  0.061463 -0.025662  0.042070 -0.072266  0.308662   
prpownr   0.300946 -0.078292  0.071566  0.336807 -0.186737  0.171352   
age      -0.052212  0.149334  0.030800  0.051650 -0.115687  0.371038   
housng    0.144170  0.059454  0.058838  0.135824 -0.077467  0.179710   
numcred   0.010728  0.408691  0.076016  0.016648 -0.056170  0.182567   
emptype   0.213519  0.024548  0.142034  0.309699 -0.388942  0.141492   
numliab  -0.021889  0.008851 -0.030063  0.003639  0.039435  0.096587   

          instrate   perstat  residlen   prpownr       age    housng  \
duration  0.073543  0.027830  0.020677  0.300946 -0.052212  0.144170   
chist     0.054341  0.028894  0.061463 -0.078292  0.149334  0.059454   
reason   -0.000763 -0.015217 -0.025662  0.071566  0.030800  0.058838   
camt     -0.264340 -0.029019  0.042070  0.336807  0.051650  0.135824   
telephne -0.022830 -0.022277 -0.072266 -0.186737 -0.115687 -0.077467   
lenemp    0.054605  0.080002  0.308662  0.171352  0.371038  0.179710   
instrate  1.000000  0.138215  0.005403  0.018823  0.018944  0.066745   
perstat   0.138215  1.000000 -0.063495 -0.006606 -0.012137  0.100881   
residlen  0.005403 -0.063495  1.000000  0.107902  0.240570 -0.027995   
prpownr   0.018823 -0.006606  0.107902  1.000000  0.041629  0.302234   
age       0.018944 -0.012137  0.240570  0.041629  1.000000  0.286900   
housng    0.066745  0.100881 -0.027995  0.302234  0.286900  1.000000   
numcred   0.049715  0.084569  0.095300 -0.019501  0.152699  0.074604   
emptype   0.109502 -0.007440 -0.011931  0.234086 -0.018429  0.092694   
numliab  -0.066688  0.122301  0.049919  0.056848  0.103699  0.139203   

           numcred   emptype   numliab  
duration  0.010728  0.213519 -0.021889  
chist     0.408691  0.024548  0.008851  
reason    0.076016  0.142034 -0.030063  
camt      0.016648  0.309699  0.003639  
telephne -0.056170 -0.388942  0.039435  
lenemp    0.182567  0.141492  0.096587  
instrate  0.049715  0.109502 -0.066688  
perstat   0.084569 -0.007440  0.122301  
residlen  0.095300 -0.011931  0.049919  
prpownr  -0.019501  0.234086  0.056848  
age       0.152699 -0.018429  0.103699  
housng    0.074604  0.092694  0.139203  
numcred   1.000000 -0.033457  0.125465  
emptype  -0.033457  1.000000 -0.079893  
numliab   0.125465 -0.079893  1.000000

Prueba estadística para determinar si nuestros predictores están relacionados: **esferacidad de Bartlet**

In [137]:
n = len(df_train); print(n)

700


In [138]:
p = df_train.shape[1];print(p)

15


In [139]:
import math as m
import scipy.stats as stats

In [140]:
mcorr=df_train.corr(method="pearson")

In [141]:
ch2=(n-1-(2*p+5/6)*m.log(np.linalg.det(mcorr)))

In [144]:
ddl=p*(p-1)/2

In [146]:
pvalue=stats.chi2.pdf(ch2,ddl);print(pvalue)

4.33466733741396e-102


In [ ]:
0.0000000000000...4334

- H nula : las variables no están correlacionadas
- H alterna: las variables están correlacionadas

- pvalor >=0.05 aceptamos la hipótesis nula
- pvalor < 0.05 rechazamos la hipótesis nula y aceptamos la alterna

Conclusión: existe un problema de multicolinealidad

**Paso 2**: ¿cuántos componentes?

**criterio de Kaiser**: varianza debe ser igual o mayor a 1

In [147]:
cov_mat = np.cov(x_train_std.T)

In [149]:
autovalores, autovectores = np.linalg.eig(cov_mat)

In [150]:
np.round(autovalores,2)

array([2.55, 1.94, 0.25, 1.29, 1.25, 1.17, 0.49, 0.53, 0.56, 0.64, 0.71,
       0.81, 0.93, 0.96, 0.95])

**Conclusión**: siguiendo el criterio de **Kaiser** el número de dimensiones es igual a 5

In [152]:
autovalores[0]/autovalores.sum()*100

17.001328383566204

**Nota**: el PC! explica el 17 % de la varianza de los datos

In [153]:
from sklearn.decomposition import PCA

In [154]:
pca = PCA(n_components=5)

In [155]:
x_train_pca = pca.fit_transform(x_train_std)

In [156]:
x_train_pca_df = pd.DataFrame(x_train_pca,
                             columns=["PC1","PC2","PC3","PC4","PC5"])

In [157]:
x_train_pca_df.head()

PC1       PC2       PC3       PC4       PC5
0 -1.187711 -1.999904  2.681628  1.523508 -1.017907
1 -3.086301  1.262873  1.443547  1.727725  0.506240
2  2.155541 -1.038378  0.147975 -0.050686 -3.117033
3 -1.456585 -0.070526 -1.667803 -0.066034  1.044369
4 -1.554560  1.322221 -0.341094  1.534566 -0.415363

In [158]:
x_train_pca_df.corr(method="pearson")

PC1           PC2           PC3           PC4           PC5
PC1  1.000000e+00  5.808671e-18  4.400003e-17  1.347269e-16 -7.366626e-17
PC2  5.808671e-18  1.000000e+00 -5.966461e-17 -5.168591e-17 -1.301291e-16
PC3  4.400003e-17 -5.966461e-17  1.000000e+00 -7.251783e-16 -2.778875e-16
PC4  1.347269e-16 -5.168591e-17 -7.251783e-16  1.000000e+00  2.077762e-16
PC5 -7.366626e-17 -1.301291e-16 -2.778875e-16  2.077762e-16  1.000000e+00

**Análisis Factorial**

In [159]:
from sklearn.decomposition import FactorAnalysis

In [160]:
fa = FactorAnalysis(n_components=5)

In [161]:
x_train_fa = fa.fit_transform(x_train_std)

In [162]:
x_train_fa_df = pd.DataFrame(x_train_fa,
                             columns=["FA1","FA2","FA3","FA4","FA5"])

In [163]:
x_train_fa_df.corr(method="pearson")

FA1           FA2           FA3           FA4           FA5
FA1  1.000000e+00 -1.067555e-13  1.086080e-13  2.389824e-14 -8.562856e-14
FA2 -1.067555e-13  1.000000e+00 -2.088004e-13  3.676634e-14 -7.101854e-14
FA3  1.086080e-13 -2.088004e-13  1.000000e+00 -2.574258e-14 -2.131784e-14
FA4  2.389824e-14  3.676634e-14 -2.574258e-14  1.000000e+00 -5.536686e-14
FA5 -8.562856e-14 -7.101854e-14 -2.131784e-14 -5.536686e-14  1.000000e+00